File: cpp/0322.0_Coin_Change.cpp
/*
DP

执行用时：88 ms, 在所有 C++ 提交中击败了38.72% 的用户
内存消耗：14.1 MB, 在所有 C++ 提交中击败了21.46% 的用户
通过测试用例：188 / 188
*/

class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        vector<unsigned int> dp(amount + 1, INT_MAX);
        dp[0] = 0;
        for (int j = 1; j <= amount; ++j) {
            for (int i = 0; i < coins.size(); ++i) {
                if (j >= coins[i])
                    dp[j] = min(dp[j], dp[j - coins[i]] + 1);
            }
        }
        if (INT_MAX == dp[amount]) {
            return -1;
        } else {
            return dp[amount];
        }
    }
};


/*
DP + memoization

执行用时：72 ms, 在所有 C++ 提交中击败了68.81% 的用户
内存消耗：15.5 MB, 在所有 C++ 提交中击败了15.77% 的用户
通过测试用例：188 / 188
*/
class Solution {
    vector<int> count;
    int dp(vector<int>& coins, int rem) {
        if (rem == 0) return 0;
        if (count[rem] != 0) return count[rem];
        int step = INT_MAX;
        for (int c: coins) {
            if (rem >= c) {
                int step1 = dp(coins, rem - c);
                if (step1 >= 0 && step1 + 1 < step)
                    step = step1 + 1;
            }
        }
        count[rem] = step == INT_MAX ? -1 : step;
        return count[rem];
    }
public:
    int coinChange(vector<int>& coins, int amount) {
        if (amount < 1) return 0;
        count.resize(amount + 1);
        return dp(coins, amount);
    }
};

File: leetcode-cn/0322.0_Coin_Change.py
'''
DP

执行用时：812 ms, 在所有 Python3 提交中击败了96.42% 的用户
内存消耗：15.3 MB, 在所有 Python3 提交中击败了30.85% 的用户
通过测试用例：188 / 188
'''
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        dp = [float('inf')] * (amount + 1)
        dp[0] = 0
        for i in range(1, amount + 1):
            dp[i] = min(dp[i - c] if i >= c else float('inf') for c in coins) + 1
        return dp[-1] if dp[-1] != float('inf') else -1

'''
DFS + memoization

执行用时：1112 ms, 在所有 Python3 提交中击败了87.78% 的用户
内存消耗：30.2 MB, 在所有 Python3 提交中击败了8.20% 的用户
通过测试用例：188 / 188
'''
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        import functools
        @functools.lru_cache(None)
        def helper(amount):
            if 0 == amount:
                return 0
            return min(helper(amount - c) if amount >= c else float('inf') for c in coins) + 1
        
        return ans if (ans := helper(amount)) != float('inf') else -1


'''
BFS
q.popleft()
q.append()

执行用时：424 ms, 在所有 Python3 提交中击败了99.14% 的用户
内存消耗：16 MB, 在所有 Python3 提交中击败了17.21% 的用户
通过测试用例：188 / 188
'''
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        q = deque([amount])
        step = 0
        visited = set()
        while q:
            for _ in range(len(q)):
                remainder = q.popleft()
                if 0 == remainder:
                    return step 
                for c in coins:
                    if remainder >= c and remainder - c not in visited:
                        visited.add(remainder - c)
                        q.append(remainder - c)
            step += 1
        return -1
        

'''
q.pop()
q.appendleft()

执行用时：424 ms, 在所有 Python3 提交中击败了99.14% 的用户
内存消耗：15.8 MB, 在所有 Python3 提交中击败了18.40% 的用户
通过测试用例：188 / 188
'''
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        q = deque([amount])
        step = 0
        visited = set()
        while q:
            for _ in range(len(q)):
                remainder = q.pop()
                if 0 == remainder:
                    return step 
                for c in coins:
                    if remainder >= c and remainder - c not in visited:
                        visited.add(remainder - c)
                        q.appendleft(remainder - c)
            step += 1
        return -1


'''
DP
完全背包问题, 无限数量

执行用时：1236 ms, 在所有 Python3 提交中击败了64.10% 的用户
内存消耗：15.1 MB, 在所有 Python3 提交中击败了86.04% 的用户
通过测试用例：188 / 188
'''
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        dp = [float('inf')] * (amount + 1)
        dp[0] = 0
        for c in coins:
            if c <= amount:
                dp[c] = 1
        
        for i in range(1, amount + 1):
            for c in coins:
                if i >= c:
                    dp[i] = min(dp[i], dp[i - c] + 1)
                    
        return dp[-1] if dp[-1] != float('inf') else -1




